sysboundary/pml.cpp

```diff
@@ -0,0 +1,275 @@
+/*
+ * This file is part of Vlasiator.
+ * Copyright 2010-2016 Finnish Meteorological Institute
+ *
+ * For details of usage, see the COPYING file and read the "Rules of the Road"
+ * at http://www.physics.helsinki.fi/vlasiator/
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+/*!\file mpl.cpp
+ * \builds the arrays needed for the PML.
+ */
+
+#include "pml.h"
+#include <cstdlib>
+#include <iostream>
+#include <iomanip> 
+#include <cmath>
+#include <sstream>
+#include <ctime>
+#include <omp.h>
+#include "fsgrid.hpp"
+#include "../definitions.h"
+#include "../common.h"
+#include "../parameters.h"
+#include "../readparameters.h"
+
+
+
+
+bool buildPMLGrids(
+    FsGrid<std::array<Real, fsgrids::pml::N_PML>, 2> &pmlGrid)
+{
+
+   typedef Parameters P;
+
+
+   /*----Get PML Array and Domain Size-----*/
+   std::array<Real, fsgrids::pml::N_PML> *pmlValue;
+   int start = 2;   /*don't judge will change it */
```

Comment:
Indeed this will not work in AMR setups.

Reply:
Suggestion: instead of this `start` and the repeated calculation of the x/y/z positions that factor into the calculation at each iteration, these positions could be stored as extra fields in the pmlGrid and computed at initialisation. This would have the added benefit that you could there make use of the determination of the layers, as this is a similar process.